### ISSUE TITLE
flakes: nix registry acts weird when adding targets that are already in the Nix store

### ISSUE THREAD

**Issue by therealpxc** [NixOS/nix#5700]
**Describe the bug**

If you add a symlink that points into the Nix store as a target for the user flake registry, when you run `nix search` you get back something like

    error: access to path '/nix/store/.../path/flake.nix' is forbidden in restricted mode
    
which seems pretty weird, since surely what is already in the Nix store must be acceptable for access in restricted mode. If you add the store path directly to the registry, things work again, but you get a message about how the store path is being copied (like into the store) and a delay before anything else happens every time you run another flake-enabled Nix command, like

    copying '/nix/store/.../path'

If the flake being copied is as big as Nixpkgs, the delay is very noticeable.

**Expected behavior**

Adding a reference (including an indirect one via symlinks) to an existing store path to the user Nix registry Just Works™ and doesn't cause any performance penalty or delay.

**`nix-env --version` output**
nix-env (Nix) 2.4

**Additional context**

This is easy to test with Nix 2.4 on NixOS if you have any leftover Nix channels. Just try

    nix registry add badnixpkgs /nix/var/nix/profiles/per-user/root/channels/nixos
    
and then

    nix search badnixpkgs whatever
    
to see the first behavior and

    nix registry add badnixpkgs $(realpath /nix/var/nix/profiles/per-user/root/channels/nixos)

then

    nix search badnixpkgs whatever

to see the second.

**Comment by regnat** [comment]
I don’t think that’s related to the Nix store. The issue is that “external” symlinks won’t work inside the registry. A simple repro that doesn’t involve any store path:

```console
$ nix flake new testFlake
$ ln -s $PWD/testFlake testSymlink        
$ nix registry add test $PWD/testSymlink        
$ nix search test hello                 
error: access to path '/tmp/tmp.h4P5AadFDo/testFlake/flake.nix' is forbidden in restricted mode
```

Lifting this restriction in general sounds a bit too much (you don’t want to traverse all your local registry entries to find symlinks arbitrarily deep inside of them), but maybe it would make sense to allow a flake registry entry to be a symlink.

**Comment by therealpxc** [comment]
Ah! Thanks for clarifying the real issue here, and for the prompt response. :)

> Lifting this restriction in general sounds a bit too much (you don’t want to traverse all your local registry entries to find symlinks arbitrarily deep inside of them),

Isn't the Nix world kinda full of (in principle) arbitrarily deep symlink forests? Plus the default behavior or commands that consult the global, user, and system Nix registries is to *hit the network* to check for updates very frequently, which is surely more expensive than symlink traversal. And for projects, can't we just resolve the symlink path when we lock a flake?

Maybe warning users about expensive traversals in the user, global, and system registries to let them know about their options for dealing with that (pinning or changing some TTL variable IIRC) would be enough. Does that make sense, or did you have other concerns in mind?

**Comment by ianthehenry** [comment]
I ran into this same behavior, and was surprised by it.

I was trying to add a flake that reflected the contents of my current channel, in order to make a bridge between the new and old worlds. But it didn't work:

```
$ nix search path:$HOME/.nix-defexpr/channels/nixpkgs/ hello
error: access to path '/nix/store/dd99fd7ik49hfc7ywiy85n2wbylhamjr-nixpkgs-22.05pre335173.56cbe42f166/nixpkgs/flake.nix' is forbidden in restricted mode
```

I didn't know what to do about that error message, and couldn't find any documentation on "restricted mode." But @tomberek pointed out that I could run this instead:

```
$ nix search path:$(realpath $HOME/.nix-defexpr/channels/nixpkgs/) hello
copying '/nix/store/dd99fd7ik49hfc7ywiy85n2wbylhamjr-nixpkgs-22.05pre335173.56c
...
```

Which does work, after a fifteen second delay.

In my head these two commands "should be" equivalent, and I don't have a sufficiently good mental model of what's going on to understand why they're treated differently -- I don't know why there are implications for deeply nested symlinks.

If they were equivalent, I could add the `nixpkgs` symlink to my registry -- adding the `realpath` sort of defeats the point of tracking my channel updates. But the fifteen second delay makes this approach pretty unattractive, so I feel like I need to find a better backwards-compatibility shim anyway.

**Comment by therealpxc** [comment]
For me, `nix search` with paths that are symiinks to channels in the Nix store only takes 2 seconds (which still feels kinda long and I don't understand why it says it's copying a store path (presumably back into the store?) every time. Do you get the 15 second delay every time, or only when Nix channel has been updated and the new channel hasn't been evaluated as a flake in this way yet? Is Nix running in impure mode for your `nix search` for some reason? When I run ```nix search --impure --no-update-lock-file path:$(realpath /nix/var/nix/profiles/per-user/root/channels/nixos) hello``` then it takes 10-15 seconds.

(I feel like performance should be int he same ballparks for us since we're both on recent-ish laptops with NVMe SSDs :shrug: )

**Comment by ianthehenry** [comment]
Yeah, the "copying" step takes about fifteen seconds for me every time. For a "fresh" evaluation, it takes about 27 seconds (15 to copy, 12 to search).

First time:

```
$ time nix search path:$(realpath ~/.nix-defexpr/channels/nixpkgs) git
copying '/nix/store/dd99fd7ik49hfc7ywiy85n2wbylhamjr-nixpkgs-22.05pre335173.56c
...
5.04s user 14.44s system 71% cpu 27.142 total
```

Next time:

```
$ time nix search path:$(realpath ~/.nix-defexpr/channels/nixpkgs) git
...
4.62s user 11.17s system 93% cpu 16.921 total
```

The 15 second estimate comes from manually timing the delay until it starts printing results with a stopwatch.

Running with `--impure` is even slower:

```
$ nix search --impure path:$(realpath $HOME/.nix-defexpr/channels/nixpkgs/) git
...
19.54s user 16.85s system 65% cpu 55.508 total
```

So I don't think that's the culprit. I'm doing my testing on a "MacBook Pro (Retina, 13-inch, Early 2015)," and I'm not sure if that qualifies as recent-ish anymore :). It does seem that it's doing actual disk I/O the whole time, for what it's worth.

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
